subsys/bluetooth/host/att.c

```diff
@@ -105,122 +125,192 @@ static void att_req_destroy(struct bt_att_req *req)
 	bt_att_req_free(req);
 }
 
-static struct bt_att *att_get(struct bt_conn *conn)
+typedef void (*bt_att_chan_sent_t)(struct bt_att_chan *chan);
+
+static bt_att_chan_sent_t chan_cb(struct net_buf *buf);
+
+void att_sent(struct bt_conn *conn, void *user_data)
 {
-	struct bt_l2cap_chan *chan;
+	struct bt_l2cap_chan *chan = user_data;
 
-	chan = bt_l2cap_le_lookup_tx_cid(conn, BT_L2CAP_CID_ATT);
-	__ASSERT(chan, "No ATT channel found");
+	BT_DBG("conn %p chan %p", conn, chan);
 
-	return CONTAINER_OF(chan, struct bt_att, chan);
+	if (chan->ops->sent) {
+		chan->ops->sent(chan);
+	}
 }
 
-static bt_conn_tx_cb_t att_cb(struct net_buf *buf);
-
-static int att_send(struct bt_conn *conn, struct net_buf *buf,
-		    bt_conn_tx_cb_t cb, void *user_data)
+static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
+		     bt_att_chan_sent_t cb)
 {
 	struct bt_att_hdr *hdr;
 
 	hdr = (void *)buf->data;
 
 	BT_DBG("code 0x%02x", hdr->code);
 
+	chan->sent = cb ? cb : chan_cb(buf);
+
+	if (IS_ENABLED(CONFIG_BT_EATT) &&
+	    atomic_test_bit(chan->flags, ATT_ENHANCED)) {
+		if (hdr->code == BT_ATT_OP_SIGNED_WRITE_CMD) {
+			return -ENOTSUP;
+		}
+
+		if (!atomic_test_bit(chan->chan.chan.status,
+				     BT_L2CAP_STATUS_OUT)) {
+			return -EAGAIN;
+		}
```

Comment:
bt_l2cap_chan_send will just queue the buf thus why the check is here so we continue attempting with another channel which may be available immediatelly, in case none can send the buf immediatelly it is put back to ATT queue.